### PR TITLE
ResponseFactory: Add a response factory to the public API

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

This change refactors and exposes a factory to create the item and feed responses. This allows users to create the responses which is necessary if users want to create a custom wrapper for the container abstraction.

```c#

ItemResponse<Book> itemResponse = this.cosmosClient.ResponseFactory.CreateItemResponse<Book>(responseMessage);
```

```c#
public abstract CosmosResponseFactory
internal abstract CosmosResponseFactoryInternal : CosmosResponseFactory
internal sealed CosmosResponseFactoryCore : CosmosResponseFactoryInternal
```

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 1]
```diff
@@ -256,6 +256,11 @@ protected CosmosClient()
         /// </summary>
         public virtual CosmosClientOptions ClientOptions => this.ClientContext.ClientOptions;
 
+        /// <summary>
+        /// The response factory used to initialize CosmosClient response types
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:260]: Lets please include more context on when to use. We can specifically call out the Encryption API.

File: Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactoryInternal.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
+//------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos
+{
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.Scripts;
+
+    internal abstract class CosmosResponseFactoryInternal : CosmosResponseFactory
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactoryInternal.cs:10]: Who is the consumer? Is it needed?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactoryInternal.cs:10]: It's just an internal abstraction to allow mocking, and it allows the implementation to be sealed.

**Comment by ealsur** [comment]
This uses the default serializer, correct? So users with a custom one would still need to do it on their end.

If it uses the default serializer, can we explain more why would you use this versus just calling the Type APIs? Sounds like it's a very specific scenario.

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 3]
```diff
@@ -256,6 +256,11 @@ protected CosmosClient()
         /// </summary>
         public virtual CosmosClientOptions ClientOptions => this.ClientContext.ClientOptions;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:257]: Is serialzier out a separate PR?

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 1 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs]: Please check the updated text.

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs (hunk 3 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:257]: I'm doing it in a separate PR.

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 4]
```diff
@@ -256,6 +256,17 @@ protected CosmosClient()
...
+        /// <summary>
+        /// The response factory used to create CosmosClient response types.
+        /// </summary>
+        /// <remarks>
+        /// This can be used for generating responses for tests, and allows users to create
+        /// a custom container that modifies the response. For example the client encryption
+        /// uses this to decrypt responses before returning to the caller.
+        /// user composition to 
+        /// </remarks>
+        public virtual CosmosResponseFactory ResponseFactory => this.ClientContext.ResponseFactory;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:268]: #preview

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 5]
```diff
@@ -256,6 +256,22 @@ protected CosmosClient()
...
         public virtual CosmosClientOptions ClientOptions => this.ClientContext.ClientOptions;
 
+        /// <summary>
+        /// The response factory used to create CosmosClient response types.
+        /// </summary>
+        /// <remarks>
+        /// This can be used for generating responses for tests, and allows users to create
+        /// a custom container that modifies the response. For example the client encryption
+        /// uses this to decrypt responses before returning to the caller.
+        /// user composition to
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/CosmosClient.cs:266]: missing text?

File: Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactory.cs [hunk 6]
```diff
@@ -4,253 +4,44 @@
...
 {
-    using System;
-    using System.IO;
-    using System.Threading.Tasks;
     using Microsoft.Azure.Cosmos.Scripts;
 
-    internal class CosmosResponseFactory
+    /// <summary>
+    /// This response factory converts response messages
+    /// to the corresponding type response.
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosResponseFactory.cs:11]: nit: "using the default serializer"?

**Pull request merged**